### PR TITLE
CRI design proposal for edged

### PR DESCRIPTION
**Pull request by gpinaik**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup

 /kind design
> /kind documentation
> /kind test
> /kind failing-test
> /kind feature

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

### CODE REVIEW

File: docs/proposals/cri.md [hunk 1]
```diff
@@ -0,0 +1,313 @@
...
+###  Edged object creation modifications
+
+The existing newEdged() function needs to modified include creating CRI runtime object based on the runtime type including
+creations of objects for runtime and image services. However the existing edged does not provide the support for all the 
+parameters required to create the CRI runtime object and default parameters need to be considered for the same.
+
+```go
+
+//newEdged creates new edged object and initialises it
+func newEdged() (*edged, error) {
```
Comment (m1093782566) [inline docs/proposals/cri.md]: I feel it's improper to give some many codes here, pseudocode instead please.

File: docs/proposals/cri.md [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+    * [Data structure modifications](#data-structure-modifications)
+    * [Edged object creation modifications](#edged-object-creation-modifications)
+    * [Runtime dependent module modifications](#runtime-dependent-module-modifications)
+    * [Runtime dependent functional modifications](#runtime-dependent-functional-modifications)
+  * [Open Questions](#open-questions) 
+    
+    
+## Motivation
+This proposal addresses the Container Runtime Interface support in edged to enable supporting multiple container
+runtimes like docker, containerd, crio etc on the edge platform.
```
Comment (m1093782566) [inline docs/proposals/cri.md]: crio -> cri-o please.

Comment (m1093782566) [reply docs/proposals/cri.md]: on the edge platform -> on the edge node.

File: docs/proposals/cri.md [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+
+### Goals
+CRI support in edged must:
+* support multiple runtimes like docker, contianerd, crio etc.
```
Comment (m1093782566) [inline docs/proposals/cri.md]: cri-o please.

File: docs/proposals/cri.md [hunk 4]
```diff
@@ -0,0 +1,313 @@
...
+* Support corresponding CNI with pause container and IP
+
+### Non-goals
+
+* Automatic detection of container runtimes and its selection.
+* 
+## Proposal
+
+Currently Kubernetes kubelet CRI supports container runtimes like containerd, crio etc but support for docker runtime is
```
Comment (m1093782566) [inline docs/proposals/cri.md]: crio -> cri-o

File: docs/proposals/cri.md [hunk 5]
```diff
@@ -0,0 +1,313 @@
...
+currently kubeedge edged supports only docker runtime using the legacy dockertools. Hence we propose to support multiple 
+container runtime in kubeedge edged as follows
+1. Include CRI support as in kubernetes kubelet to support contianerd, crio etc
+2. Continue with docker runtime support using legacy dockertools until CRI support for the same is available i.e. support
+for docker runtime using dockershim is not considered in edged
+
+
+### Use Cases
+
+* Customer can choose the required runtime on his edge platform
```
Comment (m1093782566) [inline docs/proposals/cri.md]: Please supply the use case that "in the device edge node the resource is limitted and unable to run docker daemon, therefore require a lightweighted container runtime".

File: docs/proposals/cri.md [hunk 6]
```diff
@@ -0,0 +1,313 @@
...
+
+
+## High Level Design
+
+### Edged with CRI support
+<img src="../docs/images/edged/docker-cri.png">
+
+## Low Level Design
+
+### Packages
```
Comment (m1093782566) [inline docs/proposals/cri.md]: We don't need this section.

File: docs/proposals/cri.md [hunk 7]
```diff
@@ -0,0 +1,313 @@
...
+
+Additional CRI runtime packages need to be included.
+
+```go
+"k8s.io/kubernetes/pkg/kubelet/kuberuntime"
+"k8s.io/kubernetes/pkg/kubelet/remote"
+
+```
+
+### Configuration parameters
```
Comment (m1093782566) [inline docs/proposals/cri.md]: Please list the new parameters which will be added to edge-core.

File: docs/proposals/cri.md [hunk 8]
```diff
@@ -0,0 +1,313 @@
...
+
+The options to configure runtime type i.e. docker or other remote runtimes(crio, containerd etc) and the CRI remote endpoints
+for runtime and image services should be provided.
+
+```go
+type Config struct {
```
Comment (m1093782566) [inline docs/proposals/cri.md]: We don't need to show the code-level data structure in design document unless it's API.

File: docs/proposals/cri.md [hunk 9]
```diff
@@ -0,0 +1,313 @@
...
+    e.runtime.CleanupOrphanedPod(pods)
+  ....
+}
+
+```
+
+
+## Open questions
+- Do we need to support pause container and IP in edged ?
+- Do we need to support CNI in edged ?
```
Comment (m1093782566) [inline docs/proposals/cri.md]: CNI support is out-of-scope in this document.

File: docs/proposals/cri.md [hunk 10]
```diff
@@ -0,0 +1,313 @@
...
+  if (e.containerRuntimeName == DockerContainerRuntime)
+    e.runtime.CleanupOrphanedPod(pods)
+  ....
+}
+
+```
+
+
+## Open questions
+- Do we need to support pause container and IP in edged ?
```
Comment (m1093782566) [inline docs/proposals/cri.md]: I find it's in the goals, why still need to discuss here?

File: docs/proposals/cri.md [hunk 11]
```diff
@@ -0,0 +1,313 @@
...
+
+### Use Cases
+
+* Customer can choose the required runtime on his edge platform
+
+
+## High Level Design
+
+### Edged with CRI support
+<img src="../docs/images/edged/docker-cri.png">
```
Comment (m1093782566) [inline docs/proposals/cri.md]: The image is broken.

File: docs/proposals/cri.md [hunk 12]
```diff
@@ -0,0 +1,313 @@
...
+    
+    
+## Motivation
+This proposal addresses the Container Runtime Interface support in edged to enable supporting multiple container
+runtimes like docker, containerd, crio etc on the edge platform.
+
+### Goals
+CRI support in edged must:
+* support multiple runtimes like docker, contianerd, crio etc.
+* Support corresponding CNI with pause container and IP
```
Comment (sids-b) [inline docs/proposals/cri.md]: Is CNI a goal ? We plan to support edge-to-edge communication between containers using edge-mesh.

**Comment by sids-b** [comment]
Can you please place the image in images/proposals folder ? We will soon be moving proposals to a different repository.

File: docs/proposals/cri.md (hunk 2 above)
Comment (gpinaik) [reply docs/proposals/cri.md]: Addressed

File: docs/proposals/cri.md (hunk 3 above)
Comment (gpinaik) [reply docs/proposals/cri.md]: Addressed

File: docs/proposals/cri.md (hunk 4 above)
Comment (gpinaik) [reply docs/proposals/cri.md]: Addressed

File: docs/proposals/cri.md [hunk 13]
```diff
@@ -0,0 +1,313 @@
...
+
+* Automatic detection of container runtimes and its selection.
+* 
+## Proposal
+
+Currently Kubernetes kubelet CRI supports container runtimes like containerd, crio etc but support for docker runtime is
+provided using dockershim as well. However going forward even docker runtime will be supported through only CRI. Also 
+currently kubeedge edged supports only docker runtime using the legacy dockertools. Hence we propose to support multiple 
+container runtime in kubeedge edged as follows
+1. Include CRI support as in kubernetes kubelet to support contianerd, crio etc
```
[the remark this answers is not in the document]
Comment (gpinaik) [reply docs/proposals/cri.md]: Addressed

File: docs/proposals/cri.md (hunk 5 above)
Comment (gpinaik) [reply docs/proposals/cri.md]: Addressed

File: docs/proposals/cri.md (hunk 6 above)
Comment (gpinaik) [reply docs/proposals/cri.md]: Addressed

File: docs/proposals/cri.md (hunk 7 above)
Comment (gpinaik) [reply docs/proposals/cri.md]: Addressed

File: docs/proposals/cri.md (hunk 9 above)
Comment (gpinaik) [reply docs/proposals/cri.md]: Addressed

File: docs/proposals/cri.md (hunk 10 above)
Comment (gpinaik) [reply docs/proposals/cri.md]: Addressed

File: docs/proposals/cri.md (hunk 11 above)
Comment (gpinaik) [reply docs/proposals/cri.md]: Addressed

File: docs/proposals/cri.md (hunk 12 above)
Comment (gpinaik) [reply docs/proposals/cri.md]: Addressed

File: docs/proposals/cri.md (hunk 1 above)
Comment (gpinaik) [reply docs/proposals/cri.md]: Understand but easier to specify code than pseudo code :-)

File: docs/proposals/cri.md [hunk 14]
```diff
@@ -0,0 +1,313 @@
...
+* Support corresponding CNI with pause container and IP
+
+### Non-goals
+
+* Automatic detection of container runtimes and its selection.
+* 
+## Proposal
+
+Currently Kubernetes kubelet CRI supports container runtimes like containerd, crio etc but support for docker runtime is
+provided using dockershim as well. However going forward even docker runtime will be supported through only CRI. Also
```
[the remark this answers is not in the document]
Comment (gpinaik) [reply docs/proposals/cri.md]: going forward k8s community will be removing kubelet dependency from dockershim and use CRI

File: docs/proposals/cri.md (hunk 8 above)
Comment (gpinaik) [reply docs/proposals/cri.md]: Specifying the structure as it is used below

File: docs/proposals/cri.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (CindyXing) [reply docs/proposals/cri.md]: Do we understand how community is going to change the dockershim and related? Any impact to our design?

[the remark this answers is not in the document]
Comment (gpinaik) [reply docs/proposals/cri.md]: Check kubernetes issue 73933...assume that we will have a separate cri-dockerd which should work with the existing kubelet CRI...since we are trying to incorporate the same CRI support we should not have any impact on the design

**Comment by kadisi** [comment]
can i participate this pr？？because we use containerd not docker。so i also want to kubeedge support containerd

**Comment by sids-b** [comment]
That would be great @kadisi . As far as I know @gpinaik has already started with the implementation but you guys can collaborate on remaining implementation.

@gpinaik please sync with @kadisi for collaboration on remaining effort.

**Comment by m1093782566** [comment]
@kadisi 

Definitely!

**Comment by m1093782566** [comment]
@kadisi We ALWAYS welcome any contributions. It's never too late :)

**Comment by kadisi** [comment]
@m1093782566 ok , i try it

**Comment by kadisi** [comment]
do we need use import  dockershim to support cri?

**Comment by edisonxiang** [comment]
Hi @gpinaik any updates about this pr?

File: docs/proposals/cri.md [hunk 15]
```diff
@@ -0,0 +1,308 @@
...
+## Low Level Design
+
+### Configuration parameters
+
+The following configuration parameters need to be added
+
+No | Parameter | Type | Values | Description
+---|---|---|---|---
+1 | containerRuntimeName | string | DockerRuntime/RemoteRuntime | Runtime name
+2   | RemoteRuntimeEndpoint | string | /var/run/*.sock | Endpoint of remote runtime service
```
Comment (rohitsardesai83) [inline docs/proposals/cri.md]: the formatting doesn't look correct.

**Comment by rohitsardesai83** [comment]
@gpinaik , please see if any outstanding comments need to be fixed . we need to merge the design PR.

File: docs/getting-started/release_package.md [hunk 16]
```diff
@@ -77,7 +77,12 @@
       # verify the configurations before running cloud(edgecontroller)
```
Comment (rohitsardesai83) [inline docs/getting-started/release_package.md]: the changes to release_package.md for cri should be part of a separate PR. This is the design PR.

**Comment by kevin-wangzefeng** [comment]
Please split `docs/getting-started/release_package.md` changes to a separate PR, thanks.

File: docs/proposals/cri.md (hunk 1 above)
Comment (gpinaik) [reply docs/proposals/cri.md]: Used pseudocode

File: docs/getting-started/release_package.md (hunk 16 above)
Comment (gpinaik) [reply docs/getting-started/release_package.md]: Removed

**Comment by rohitsardesai83** [comment]
Now we do support docker runtime via dockershim. When this proposal was written, this support was not there. The proposal needs to be modified to reflect these changes ? Or we merge this proposal and then amend it with the latest changes ?

**Comment by kevin-wangzefeng** [comment]
We can get it accepted first, and append updates with another PR.

**Comment by sids-b** [comment]
squashing and merging this one.
